contracts/ronin/validator/CoinbaseExecution.sol

```diff
@@ -217,31 +218,40 @@ abstract contract CoinbaseExecution is
   function _distributeRewardToTreasuriesAndCalculateTotalDelegatingReward(
     uint256 lastPeriod,
     address[] memory currValidatorIds
-  ) private returns (uint256 totalDelegatingReward, uint256[] memory delegatingRewards) {
+  )
+    private
+    returns (uint256 totalDelegatingReward, uint256[] memory delegatingRewards, uint256[] memory delegatingFFRewards)
+  {
     address vId; // validator id
     address payable treasury;
-    delegatingRewards = new uint256[](currValidatorIds.length);
 
-    for (uint _i; _i < currValidatorIds.length;) {
-      vId = currValidatorIds[_i];
+    uint256 length = currValidatorIds.length;
+    delegatingRewards = new uint256[](length);
+    delegatingFFRewards = new uint256[](length);
+
+    for (uint i; i < length; ++i) {
+      vId = currValidatorIds[i];
       treasury = _candidateInfo[vId].__shadowedTreasury;
 
       if (!_isJailedById(vId) && !_miningRewardDeprecatedById(vId, lastPeriod)) {
-        totalDelegatingReward += _delegatingReward[vId];
-        delegatingRewards[_i] = _delegatingReward[vId];
+        (uint256 validatorFFReward, uint256 delegatingFFReward) =
+          _splitRewardBetweenValidatorAndDelegators({ vId: vId, totalReward: _fastFinalityReward[vId] });
+
+        delegatingFFRewards[i] = delegatingFFReward;
+        // Add the fast finality reward to the total delegating reward array
+        delegatingRewards[i] = _delegatingReward[vId] + delegatingFFReward;
+
+        totalDelegatingReward += delegatingRewards[i];
```

Comment:
Check with Thuc Ho to consider removing this totalDelegatingReward. Two events of delegating reward are currently inconsistent.